crates/redis/src/lib.rs

```diff
@@ -86,56 +138,34 @@ impl TriggerExecutor for RedisTrigger {
-            pubsub.subscribe(channel).await?;
-            println!("\t{channel}: [{}]", component.join(","));
-        }
+        if let Some(channels) = self.server_channels.get(&address) {
+            for (channel, component) in channels.iter() {
+                tracing::info!("Subscribing component {component:?} to channel {channel:?}");
+                pubsub.subscribe(channel).await?;
+                println!("\t{address}:{channel}: [{}]", component.join(","));
+            }
 
-        let mut stream = pubsub.on_message();
-        loop {
-            match stream.next().await {
-                Some(msg) => {
-                    if let Err(err) = self.handle(msg).await {
-                        tracing::warn!("Error handling message: {err}");
+            let mut stream = pubsub.on_message();
+            loop {
+                match stream.next().await {
+                    Some(msg) => {
+                        if let Err(err) = self.handle(address.clone(), msg).await {
+                            tracing::warn!("Error handling message: {err}");
+                        }
                     }
-                }
-                None => {
-                    tracing::trace!("Empty message");
-                    if !client.check_connection() {
-                        tracing::info!("No Redis connection available");
-                        break Ok(());
+                    None => {
+                        tracing::trace!("Empty message");
+                        if !client.check_connection() {
+                            tracing::info!("No Redis connection available");
+                            println!("Disconnected from {address}");
+                            break;
+                        }
                     }
-                }
-            };
-        }
-    }
-}
-
-impl RedisTrigger {
-    // Handle the message.
-    async fn handle(&self, msg: redis::Msg) -> Result<()> {
-        let channel = msg.get_channel_name();
-        tracing::info!("Received message on channel {:?}", channel);
-
-        if let Some(component_ids) = self.channel_components.get(channel) {
-            let futures = component_ids.iter().map(|id| {
-                tracing::trace!("Executing Redis component {id:?}");
-                SpinRedisExecutor.execute(&self.engine, id, channel, msg.get_payload_bytes())
-            });
-            let results: Vec<_> = join_all(futures).await.into_iter().collect();
-            let errors = results
-                .into_iter()
-                .filter_map(|r| r.err())
-                .collect::<Vec<_>>();
-            if !errors.is_empty() {
-                return Err(anyhow!("{errors:#?}"));
+                };
             }
-        } else {
-            tracing::debug!("No subscription found for {:?}", channel);
+            return Ok(());
         }
-
-        Ok(())
+        bail!("Server does not have any channels that need to be subscribed to")
```

Comment:
This should never happen.  If it does it means your `addresses` and `server_channels` collections are out of sync.  Is it possible to design your data structure so that this cannot happen?  For example, in the `server_channels` collection, could you use the keys of that collection as your list of addresses, instead of maintaining a separate collection?

Reply:
Yes, that makes sense.